### PR TITLE
Add a new escaping '(61)' for '=' because the original one '&equals;'…

### PR DESCRIPTION
**Pull request by nwangtw**
… is not friendly to bash

### CODE REVIEW

**Comment by jerrypeng** [comment]
@nwangtw thanks for looking into this for me!  My only question is why are we replacing "=" with something else to begin with.  Its already base64 encoded and in quotes.

**Comment by nwangtw** [comment]
My guess is that "=" is a valid character in base64, and it might mess up configs potentially because configurations are in config=value format.

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java [hunk 1]
```diff
@@ -442,21 +442,24 @@ public static boolean setSchedulerLocation(
...
+   * reason
    *
    * @return encoded string
    */
   public static String encodeJavaOpts(String javaOpts) {
     String javaOptsBase64 = DatatypeConverter.printBase64Binary(
         javaOpts.getBytes(StandardCharsets.UTF_8));
-
-    return String.format("\"%s\"", javaOptsBase64.replace("=", "&equals;"));
+    return String.format("\"%s\"", javaOptsBase64.replace("=", "(61)").replace("=", "&equals;"));
```
Comment (jerrypeng) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: do we need .replace("=", "&equals;")) here?  All the "=" would already be replaced by the previous replace statement right?

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java [hunk 2]
```diff
@@ -442,21 +442,24 @@ public static boolean setSchedulerLocation(
    * Encode the JVM options
    * <br> 1. Convert it into Base64 format
    * <br> 2. Add \" at the start and at the end
-   * <br> 3. replace "=" with "&amp;equals;"
+   * <br> 3. replace "=" with "(61)" and "&amp;equals;"
+   * Note that '=' is escaped in two different ways. '(61)' is the new escaping.
+   * '&amp;equals;' was the original way but it is not friendly to bash and is
+   * causing issues in some schedulers. It is still left there for backward compatibility
```
Comment (jerrypeng) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: can we add a comment on why we need to replace "=" for future reference

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java (hunk 1 above)
Comment (nwangtw) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: right. Technically it is not really necessary.

I left it here to be consistent with the decoding path and easier (just a little) for people to understand. In future we can remove both of them at the same time.

Comment (jerrypeng) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: gotcha can we just add some comments at the places this is done for future reference.  Thanks!

Comment (jerrypeng) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: or maybe a TODO for the future

**Comment by jerrypeng** [comment]
LGTM just some minor comments

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java (hunk 2 above)
Comment (nwangtw) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: Sure. added.

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java (hunk 1 above)
Comment (nwangtw) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: Thinking about it again. Now I don't think leaving the encoding part here is very helpful. Plus it looks wrong and could cause more confusion. I am removing this replace() call in the new update.

It is also easier to tell that the decode() part is for backward compatibility purpose only without the encode() part.

**Pull request merged**